pkg/controller/handlers/workflowstep/loop.go

```diff
@@ -146,66 +137,80 @@ func elementPrompt(element, prompt string) string {
 	var checkStep v1.WorkflowStep
 	if err := client.Get(ctx, router.Key(dataStep.Namespace, dataStep.Name), &checkStep); apierrors.IsNotFound(err) {
-		return "", "", true, nil
+		return "", nil, true, nil
 	} else if err != nil {
-		return "", "", false, err
+		return "", nil, false, err
 	}
 
 	if checkStep.Status.State != types.WorkflowStateComplete || checkStep.Status.LastRunName == "" {
-		return "", "", true, nil
+		return "", nil, true, nil
 	}
 
 	var run v1.Run
-	if err := client.Get(ctx, router.Key(dataStep.Namespace, checkStep.Status.LastRunName), &run); err != nil {
-		return "", "", false, err
+	if err := client.Get(ctx, router.Key(checkStep.Namespace, checkStep.Status.LastRunName), &run); err != nil {
+		return "", nil, false, err
 	}
 
-	datasetID = getDatasetID(run.Status.Output)
-	if datasetID == "" {
-		parentStep.Status.Error = "no dataset ID found in output"
-		parentStep.Status.State = types.WorkflowStateError
-		return "", "", true, nil
+	var thread v1.Thread
+	if err := client.Get(ctx, router.Key(run.Namespace, run.Spec.ThreadName), &thread); err != nil {
+		return "", nil, false, err
 	}
 
-	return run.Name, datasetID, false, nil
-}
+	content, err := h.gptscriptClient.ReadFileInWorkspace(ctx, fileName, gptscript.ReadFileInWorkspaceOptions{
+		WorkspaceID: thread.Status.WorkspaceID,
+	})
+	if err != nil {
+		return "", nil, false, err
+	}
 
-func getDatasetID(output string) string {
-	return datasetIDRegex.FindString(output)
-}
+	if isDatasetID(string(content)) {
+		// We use the dataset package rather than making SDK calls because it is more direct and more performant.
+		// All that the SDK calls do is call out to a daemon tool that runs the same library code that we are referencing here.
+		datasetManager, err := dataset.NewManager(thread.Status.WorkspaceID)
+		if err != nil {
+			return "", nil, false, err
+		}
 
-func getWorkspaceID(ctx context.Context, client kclient.Client, step *v1.WorkflowStep) (string, error) {
-	var workflowExecution v1.WorkflowExecution
-	if err := client.Get(ctx, router.Key(step.Namespace, step.Spec.WorkflowExecutionName), &workflowExecution); err != nil {
-		return "", err
-	}
+		dataset, err := datasetManager.GetDataset(ctx, string(content))
+		if err != nil {
+			return "", nil, false, err
+		}
 
-	var thread v1.Thread
-	if err := client.Get(ctx, router.Key(step.Namespace, workflowExecution.Status.ThreadName), &thread); err != nil {
-		return "", err
+		for _, element := range dataset.GetAllElements() {
+			data = append(data, fmt.Sprintf("%s: %s", element.Name, element.Contents))
+		}
+	} else {
+		if err := json.Unmarshal(content, &data); err != nil {
+			return "", nil, false, err
+		}
 	}
 
-	return thread.Status.WorkspaceID, nil
+	return thread.Status.WorkspaceID, data, false, nil
+}
+
+func isDatasetID(output string) bool {
+	return datasetIDRegex.MatchString(output)
```

Comment:
nit: if `datasetIDRegex.Match` is used here, then you can pass the `[]byte` instead of casting to `string`.

Reply:
fixed